File: internal/lifecycle/reconcile.go
package lifecycle

import (
	"context"
	"fmt"

	"github.com/abbyfluoroethane/bento/internal/hypervisor"
	"github.com/abbyfluoroethane/bento/internal/types"
)

// ReconcileReport is the disagreement between libvirt and the database
// (SPEC 6.1). The operator reads it and corrects by hand; Bento does not.
type ReconcileReport struct {
	// DomainsWithoutRows are libvirt domains with no instances row.
	DomainsWithoutRows []hypervisor.DomainInfo
	// RowsWithoutDomains are instances rows with no libvirt domain.
	RowsWithoutDomains []types.Instance
}

// Empty reports agreement: every domain has a row and every row a domain.
func (r ReconcileReport) Empty() bool {
	return len(r.DomainsWithoutRows) == 0 && len(r.RowsWithoutDomains) == 0
}

// Reconcile compares every libvirt domain against every instances row and
// reports the differences, matched by UUID (SPEC 6.1).
//
// Reconcile reports and never deletes. It performs no write of any kind: a
// reconciliation bug that deletes a domain is worse than a row that is
// wrong.
func (m *Manager) Reconcile(ctx context.Context) (ReconcileReport, error) {
	var report ReconcileReport
	domains, err := m.hyp.List(ctx)
	if err != nil {
		return report, fmt.Errorf("lifecycle: reconcile: list domains: %w", err)
	}
	rows, err := m.store.Instances()
	if err != nil {
		return report, fmt.Errorf("lifecycle: reconcile: list instances: %w", err)
	}
	byUUID := make(map[string]bool, len(rows))
	for _, inst := range rows {
		byUUID[inst.UUID] = true
	}
	domainUUIDs := make(map[string]bool, len(domains))
	for _, dom := range domains {
		domainUUIDs[dom.UUID] = true
		if !byUUID[dom.UUID] {
			report.DomainsWithoutRows = append(report.DomainsWithoutRows, dom)
		}
	}
	for _, inst := range rows {
		if !domainUUIDs[inst.UUID] {
			report.RowsWithoutDomains = append(report.RowsWithoutDomains, inst)
		}
	}
	return report, nil
}

File: crates/lifecycle/src/reconcile.rs
use std::collections::HashSet;

use bento_hypervisor::DomainInfo;
use bento_types::Instance;

use crate::{Error, Manager, Result};

/// The disagreement between libvirt and the database (SPEC 6.1). Operators
/// inspect and correct it by hand.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ReconcileReport {
    pub domains_without_rows: Vec<DomainInfo>,
    pub rows_without_domains: Vec<Instance>,
}

impl ReconcileReport {
    pub fn is_empty(&self) -> bool {
        self.domains_without_rows.is_empty() && self.rows_without_domains.is_empty()
    }
}

impl Manager {
    /// Compares domains and rows by UUID and reports differences (SPEC 6.1).
    /// It never writes or deletes: a reconciliation bug that deletes a domain
    /// is worse than a wrong row.
    pub async fn reconcile(&self) -> Result<ReconcileReport> {
        let domains = self.hyp.list().await.map_err(|error| {
            Error::operation(format!("lifecycle: reconcile: list domains: {error}"))
        })?;
        let rows = self.store.instances().await.map_err(|error| {
            Error::operation(format!("lifecycle: reconcile: list instances: {error}"))
        })?;
        let row_ids: HashSet<_> = rows.iter().map(|row| row.uuid.clone()).collect();
        let domain_ids: HashSet<_> = domains.iter().map(|domain| domain.uuid.clone()).collect();
        Ok(ReconcileReport {
            domains_without_rows: domains
                .into_iter()
                .filter(|domain| !row_ids.contains(domain.uuid.as_str()))
                .collect(),
            rows_without_domains: rows
                .into_iter()
                .filter(|row| !domain_ids.contains(row.uuid.as_str()))
                .collect(),
        })
    }
}
